lib/components/admin-command.js

```diff
@@ -0,0 +1,157 @@
+            }
+
+            names.forEach((a) => this.optaliases[a] = name);
+        });
+    }
+}
+
+/**
+ * Invoke the handler function for the command. The given list of strings will
+ * be parsed for optional arguments as per the command's opts specification, if
+ * present.
+ * @param {Object} bridge The top-level bridge instance to pass to the handler
+ * function.
+ * @param {string[]} args A list of string arguments parsed from the command
+ * line.
+ * @param {function} respond The responder function to pass to the handler
+ * function.
+ * @return {Promise} Returns the void promise returned by the handler function.
+ */
+AdminCommand.prototype.run = function(bridge, args, respond) {
+    var opts;
+    if (this.optspec) {
+        opts = minimist(args);
+
+        // Canonicalise aliases
+        Object.keys(this.optaliases).forEach((a) => {
+            if (a in opts) {
+                opts[this.optaliases[a]] = opts[a];
+                delete opts[a];
+            }
+        });
+
+        Object.keys(opts).forEach((n) => {
+            if (n === "_") return;
+
+            if (!(n in this.optspec)) {
+                throw Error("Unrecognised option: " + n);
+            }
+        });
+
+        var missing = [];
+        Object.keys(this.optspec).sort().forEach((n) => {
+            if (this.optspec[n].required && !(n in opts)) missing.push(n);
+        });
+
+        if (missing.length) {
+            throw Error("Missing required options: " + missing.join(", "));
+        }
+
+        args = opts._;
+        delete opts["_"];
+    }
+
+    return this._func(bridge, opts, args, respond);
+};
+
+/**
+ * Helper method to construct a 'help' AdminCommand instance
+ * @param {Object} commands The object containing the application's
+ * AdminCommand instances.
+ * @return {AdminCommand} A new AdminCommand.
+ *
+ * Typically this method would be invoked just after the admin command object
+ * is declared, to provide the automatic online help command.
+ *
+ *   var commands = {};
+ *   commands.help = AdminCommand.makeHelpCommand(commands);
+ */
+AdminCommand.makeHelpCommand = function(commands) {
+    return new AdminCommand({
+        desc: "display a list of commands",
+        func: function(bridge, _, args, respond) {
+            if (args.length == 0) {
+                Object.keys(commands).sort().forEach(function (k) {
+                    var cmd = commands[k];
+                    respond(k + ": " + cmd.desc);
+                });
+            }
+            else {
+                var name = args.shift();
+                var cmd = commands[name];
+                if (!cmd) {
+                    throw Error("No such command '" + name + "'");
+                }
+
+                respond(name + " - " + cmd.desc);
+                var optspec = cmd.optspec || {};
+                Object.keys(optspec).sort().forEach((n) => {
+                    respond("  --" + n + ": " + optspec[n].desc);
```

Comment:
I'm not sure how I feel about this `respond` stuff. I assume this is a line-by-line thing, but I dislike it because the naive form is error prone:

```js
function respond(someString) {
  doHttpHitWithMessage(someString).then(function(res) {
    // yay it worked
  }, function(err) {
    // sadsies it no worky.
  });
}
```

Because `doHttpHitWithMessage` is async, this will race and you might get out-of-order messages. If we expect the `respond` implementation to be sending stuff to Matrix, this also will be grossly counting against our rate limiting count, whereas a `\n` message wouldn't.